circus-train-gcp/src/test/java/com/hotels/bdp/circustrain/gcp/GCPCredentialCopierTest.java

```diff
@@ -16,59 +16,128 @@
   public void copyCredentialsWhenFileDoesntExistThrowsException() throws Exception {
-    doNothing().when(fs).copyFromLocalFile(any(Path.class), any(Path.class));
-    String credentialProvider = "test.json";
-    GCPCredentialCopier copier = new GCPCredentialCopier(fs, conf, credentialProvider);
+    doThrow(new IOException("foo"))
+        .when(fileSystem)
+        .copyFromLocalFile(eq(new Path(credentialsFileRelativePath)), any(Path.class));
+    setGcpSecurity(credentialsFilePath, null);
+    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
+    copier.copyCredentials();
+  }
+
+  @Test(expected = IllegalArgumentException.class)
+  public void nullGCPSecurityThrowsException() throws Exception {
+    doNothing().when(fileSystem).copyFromLocalFile(any(Path.class), any(Path.class));
+    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, null);
+    copier.copyCredentials();
+  }
+
+  @Test
+  public void distributedFileSystemProvided() throws Exception {
+    doNothing().when(fileSystem).copyFromLocalFile(any(Path.class), any(Path.class));
+    temporaryFolder.newFile(credentialsFilePath);
+    setGcpSecurity(credentialsFilePath, distributedFileSystem);
+    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
+    ArgumentCaptor<Path> pathCaptor = ArgumentCaptor.forClass(Path.class);
+    copier.copyCredentials();
+    verify(fileSystem).copyFromLocalFile(any(Path.class), pathCaptor.capture());
+    Path destination = pathCaptor.getValue();
+    assertTrue(destination.toString().startsWith(distributedFileSystem));
+  }
+
+  @Test
+  public void noDistributedFileSystemProvided() throws Exception {
+    doNothing().when(fileSystem).copyFromLocalFile(any(Path.class), any(Path.class));
+    credentialsFilePath = temporaryFolder.getRoot() + credentialsFilePath;
+    temporaryFolder.newFile(credentialsFilePath);
+    setGcpSecurity(credentialsFilePath, null);
+    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
+    ArgumentCaptor<Path> pathCaptor = ArgumentCaptor.forClass(Path.class);
+    copier.copyCredentials();
+    verify(fileSystem).copyFromLocalFile(any(Path.class), pathCaptor.capture());
+    Path destination = pathCaptor.getValue();
+    assertTrue(destination.toString().startsWith(defaultDistributedFileSystemRoot));
+  }
+
+  @Test
+  public void shouldGetRelativePath() throws IOException {
+    temporaryFolder.newFile(credentialsFilePath);
+    setGcpSecurity(credentialsFilePath, null);
+    GCPCredentialCopier copier = new GCPCredentialCopier(fileSystem, conf, gcpSecurity);
+    ArgumentCaptor<Path> pathCaptor = ArgumentCaptor.forClass(Path.class);
     copier.copyCredentials();
+    verify(fileSystem).copyFromLocalFile(pathCaptor.capture(), any(Path.class));
+    Path source = pathCaptor.getValue();
+    assertTrue(source.toString().equals(credentialsFileRelativePath));
```

Comment:
This give you very bad error messages if it fails prefer: `assertThat(source.toString(), is(credentialsFileRelativePath));`